monkey/tests/unit_tests/infection_monkey/exploit/tools/test_brute_force_credentials_generator.py

```diff
@@ -0,0 +1,208 @@
+from typing import Callable, Iterable, Set
+
+from tests.data_for_tests.propagation_credentials import IDENTITIES, SECRETS
+
+from common.credentials import Credentials, Identity, LMHash, NTHash, Password, Secret
+from infection_monkey.exploit.tools import generate_brute_force_credentials
+
+
+def generate_and_compare_credentials(
+    input_credentials: Iterable[Credentials],
+    expected_credentials: Set[Credentials],
+    identity_filters: Iterable[Callable[[Identity], bool]] = (),
+    secret_filters: Iterable[Callable[[Secret], bool]] = (),
+):
+    generated_credentials = generate_brute_force_credentials(
+        input_credentials, identity_filters=identity_filters, secret_filters=secret_filters
+    )
+
+    assert len(generated_credentials) == len(expected_credentials)
+    assert set(generated_credentials) == expected_credentials
+
+
+def test_generate_all_combinations_1():
+    input_credentials = [
+        Credentials(identity=IDENTITIES[0], secret=None),
+        Credentials(identity=IDENTITIES[2], secret=None),
+        Credentials(identity=None, secret=SECRETS[0]),
+        Credentials(identity=None, secret=SECRETS[3]),
+    ]
+    expected_credentials = {
+        Credentials(identity=IDENTITIES[0], secret=SECRETS[0]),
+        Credentials(identity=IDENTITIES[0], secret=SECRETS[3]),
+        Credentials(identity=IDENTITIES[2], secret=SECRETS[0]),
+        Credentials(identity=IDENTITIES[2], secret=SECRETS[3]),
+    }
+
+    generate_and_compare_credentials(input_credentials, expected_credentials)
+
+
+def test_generate_all_combinations_2():
+    input_credentials = [
+        Credentials(identity=IDENTITIES[0], secret=SECRETS[0]),
+        Credentials(identity=IDENTITIES[2], secret=SECRETS[3]),
+    ]
+    expected_credentials = {
+        Credentials(identity=IDENTITIES[0], secret=SECRETS[0]),
+        Credentials(identity=IDENTITIES[0], secret=SECRETS[3]),
```

Comment:
I would expect second credentials to be `Credentials(identity=IDENTITIES[2], secret=SECRETS[3])` as they are not a random combination of identities and secrets.

Reply:
https://github.com/guardicore/monkey/pull/3048#discussion_r1126340588